lib/plugin.lazy.js

```diff
@@ -0,0 +1,121 @@
+  })
+
+  window.addEventListener('error', SentryMock.captureException)
+
+  // Inject Sentry mock to the context as $sentry
+  tempInject(ctx, 'sentry', SentryMock)
+  <% } %>
+
+  const loadSentryHook = () => !loaded && loadSentry(ctx, inject)
+
+  <% if (options.lazy.injectLoadHook) { %>
+  inject('sentryLoad', loadSentryHook)
+  ctx.$sentryLoad = loadSentryHook
+  <% } else { %>
+  window.onNuxtReady(loadSentryHook)
+  <% } %>
+
+  sentryReadyPromise = new Promise(resolve => {
+    sentryReadyResolve = resolve
+  })
+
+  const sentryReady = () => sentryReadyPromise
+
+  inject('sentryReady', sentryReady)
+  ctx.$sentryReady = sentryReady
+}
+
+async function loadSentry(ctx, inject) {
+  <%
+  const magicComments = [`webpackChunkName: '${options.lazy.chunkName}'`]
+  if (options.lazy.webpackPrefetch) {
+    magicComments.push('webpackPrefetch: true')
+  }
+  if (options.lazy.webpackPreload) {
+    magicComments.push('webpackPreload: true')
+  }
+  %>
+  const Sentry = await import(/* <%= magicComments.join(', ') %> */ '@sentry/browser')
+  const { <%= Object.keys(options.integrations).join(', ') %> } = await import(/* <%= magicComments.join(', ') %> */ '@sentry/integrations')
+
+  <% if (options.initialize) { %>// Initialize sentry
+  const config = <%= serialize(options.config) %>
+  config.integrations = [
+    <%= Object.entries(options.integrations).map(([name, integration]) => {
+      if (name === 'Vue') {
+        return `new ${name}({Vue: VueLib, ...${serialize(integration)}})`
+      }
+
+      const integrationOptions = Object.entries(integration).map(([key, option]) => 
+        typeof option === 'function'
+        ? `${key}:${serializeFunction(option)}`
+        : `${key}:${serialize(option)}`
+      )
+
+      return `new ${name}({${integrationOptions.join(',')}})`
+    }).join(',\n      ')%>
+  ]
+
+  Sentry.init(config)
+  <% } %>
+
+  <% if (options.lazy.injectMock) { %>
+  window.removeEventListener('error', SentryMock.captureException)
+
+  delayedCalls.forEach(([methodName, args]) => Sentry[methodName].apply(Sentry, args))
+  <% } %>
+
+  delete ctx.$sentry
+  inject('sentry', Sentry)
+
+  sentryReadyResolve(Sentry)
+  sentryReadyPromise = Promise.resolve(Sentry)
+  loaded = true
+  
+  // help gc
+  <% if (options.lazy.injectMock) { %>
+  // Dont unset delayedCalls & SentryMock during
+  // development, this will cause HMR issues
+  <% if (!options.dev) { %>
+  delayedCalls = undefined
+  SentryMock = undefined
+  <% } else { %>
+  delayedCalls = []
+  <% } %>
+  <% } %>
+  sentryReadyResolve = undefined
+}
+
+// Temporary inject into Nuxt without using Nuxt's inject method,
+// so we can later inject the final value
+// This method is adapted from the inject method in nuxt/vue-app/template/index.js
+function tempInject({ app }, key, value) {
+  key = '$' + key
+  // Add into app
+  app[key] = value
+  // Add into context
+  if (!app.context[key]) {
+    app.context[key] = value
+  }
```

Comment:
What about this part of the original inject: https://github.com/nuxt/nuxt.js/blob/6ab65fbf4f2b0d2348363c43f5c2fe81ef47cbc6/packages/vue-app/template/index.js#L191-L199

This `tempInject` doesn't inject the getter to Vue prototype. Seems like a bit of a problem?

Can we instead do something like:

```js
function customInject({ app }, originalInject, key, value) {
  key = '$' + key
  delete app.context[key]
  originalInject(key, value)
}
```

Reply:
It does, it just does it without calling a plugin: https://github.com/nuxt-community/sentry-module/blob/1cd8da3648c775322d659a69d580c882dbd311aa/lib/plugin.lazy.js#L119 Afaik using the plugin is only useful as a preventive measure for only setting the property once.

You cant call Nuxt's inject twice because the property is added using defineProperty without configurable set so you after defining it once you wont be allowed to change it anymore.

Reply:
The getter that original `inject` creates returns `this.$root.$options[key]` so it's fine if it's just added once, on first inject. The getter will return the most recent value assigned to `app[key]` (I assume that this is what `this.$root.$options[key]` refers to. So I think that my suggestion still makes sense to use both for the temp and later inject (or just the later one).

Reply:
My point is to have smaller and more concise code and also reusing as much as possible of original inject. It seems that we really only need to force-inject key to `app.context`. Otherwise the original `inject` should all the rest of the work.

Reply:
@rchl Could you please try to change this yourself maybe and make sure that it works? You are probably perfectly right but I dont have time to keep looking into finding the ultimate perfect solution unfortunately ;)

Reply:
Not immediately but yes, I can.

Reply:
It's actually quite hard to override keys injected to the Vue instance since what happens now is that those are injected to `app` object and then Vue instance is created with those options passed in and then `app` keys are copied into the instance's `$options` object. So changing Vue.prototype later will not affect those options.

Reply:
See my latest commits, this should work (it does with `yarn nuxt test/fixture/with-lazy-config/`). In theory it would be possible that Sentry is loaded at a later time then requested, because if you call sentryLoad before App.create then you have to wait until window.onNuxtReady. I added a warning about that in dev mode, but I tried and I couldnt even trigger that warning. Also it doesnt make any sense at all, because if you need that then you just shouldnt use lazy mode anyway. Therefore I think this edge-case is covered enough.